src/net_processing.cpp

```diff
@@ -1651,6 +1672,20 @@ void PeerManagerImpl::FinalizeNode(const CNode& node)
     LogPrint(BCLog::NET, "Cleared nodestate for peer=%d\n", nodeid);
 }
 
+bool PeerManagerImpl::HasAllDesirableServiceFlags(ServiceFlags services) const
+{
+    // Shortcut for (services & GetDesirableServiceFlags(services)) == GetDesirableServiceFlags(services)
+    return !(GetDesirableServiceFlags(services) & (~services));
+}
+
+ServiceFlags PeerManagerImpl::GetDesirableServiceFlags(ServiceFlags services) const
+{
+    if (services & NODE_NETWORK_LIMITED) {
+        if (m_allow_limited_peers_conn) return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
```

Comment:
I like how the best block's time is cached in `PeerManagerImpl::m_best_block_time` but further caching the expression "is the best block time older than 288 blocks" seems excessive and unnecessary to me. It makes the code more complex and obviously gets stale one second after it is set.

Here you can use something like:
```cpp
if (ApproximateBestBlockAgeInNumberOfBlocks() < NODE_NETWORK_LIMITED_MIN_BLOCKS) {
    return ServiceFlags(NODE_NETWORK_LIMITED | NODE_WITNESS);
}
...
int64_t ApproximateBestBlockAgeInNumberOfBlocks() {
    return (GetTime<std::chrono::seconds>() - m_best_block_time.load()) / m_chainparams.GetConsensus().nPowTargetSpacing;
}
```

Then the following hunks can be dropped from commit 5b4e031b437922735ff2676fa90fbf3d404ab44f `net: peer manager, dynamically adjust desirable services flag`:

```diff
@@ -787,12 +787,15 @@ private:
     /** Number of preferable block download peers. */
     int m_num_preferred_download_peers GUARDED_BY(cs_main){0};
     
     /** Stalling timeout for blocks in IBD */
     std::atomic<std::chrono::seconds> m_block_stalling_timeout{BLOCK_STALLING_TIMEOUT_DEFAULT};
 
+    /** Whether limited peers connections are accepted or not */
+    std::atomic<bool> m_allow_limited_peers_conn{false};
+
     bool AlreadyHaveTx(const GenTxid& gtxid)
         EXCLUSIVE_LOCKS_REQUIRED(cs_main, !m_recent_confirmed_transactions_mutex);
     
     /**
      * Filter for transactions that were recently rejected by the mempool.
      * These are not rerequested until the chain tip changes, at which point
@@ -1072,12 +1075,22 @@ private:
      *            False if address relay is disallowed
      */
     bool SetupAddressRelay(const CNode& node, Peer& peer) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
     
     void AddAddressKnown(Peer& peer, const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
     void PushAddress(Peer& peer, const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
+
+    /**
+     * Updates the contextual information will be used within the connections
+     * desirable services flags decision-making process.
+     */
+    void MaybeUpdateDesirableConnOptions() {
+        // If tip is further than NODE_NETWORK_LIMITED_MIN_BLOCKS, we are really behind and shouldn't connect to limit
+        std::chrono::seconds time_between_blocks{m_chainparams.GetConsensus().nPowTargetSpacing};
+        m_allow_limited_peers_conn = m_best_block_time.load() > GetTime<std::chrono::seconds>() - time_between_blocks 
+    }
 };
 
 const CNodeState* PeerManagerImpl::State(NodeId pnode) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
     std::map<NodeId, CNodeState>::const_iterator it = m_node_states.find(pnode);
     if (it == m_node_states.end())
@@ -2068,12 +2080,13 @@ void PeerManagerImpl::NewPoWValidBlock(const CBlockIndex *pindex, const std::sha
  * Update our best height and announce any block hashes which weren't previously
  * in m_chainman.ActiveChain() to our peers.
  */
 void PeerManagerImpl::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownlo
 {
     SetBestBlock(pindexNew->nHeight, pindexNew->GetBlockTime());
+    MaybeUpdateDesirableConnOptions();
     SetServiceFlagsIBDCache(!fInitialDownload);
     
     // Don't relay inventory during initial block download.
     if (fInitialDownload) return;
     
     // Find the hashes of all blocks that weren't previously in the best chain.
@@ -5272,21 +5285,28 @@ void PeerManagerImpl::CheckForStaleTipAndEvictPeers()
     
     auto now{GetTime<std::chrono::seconds>()};
     
     EvictExtraOutboundPeers(now);
     
     if (now > m_stale_tip_check_time) {
-        // Check whether our tip is stale, and if so, allow using an extra
-        // outbound peer
-        if (!m_chainman.m_blockman.LoadingBlocks() && m_connman.GetNetworkActive() && m_connman.GetUseAddrmanOutgoing(
-            LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d seconds a
-                      count_seconds(now - m_last_tip_update.load()));
-            m_connman.SetTryNewOutboundPeer(true);
-        } else if (m_connman.GetTryNewOutboundPeer()) {
-            m_connman.SetTryNewOutboundPeer(false);
+        // Skip checks when blocks are being loaded from disk or the network is disabled
+        if (!m_chainman.m_blockman.LoadingBlocks() && m_connman.GetNetworkActive()) {
+            // Check whether our tip is stale, and if so, allow using an extra outbound peer
+            if (m_connman.GetUseAddrmanOutgoing() && TipMayBeStale()) {
+                LogPrintf("Potential stale tip detected, will try using extra outbound peer (last tip update: %d secon
+                          count_seconds(now - m_last_tip_update.load()));
+                m_connman.SetTryNewOutboundPeer(true);
+            } else if (m_connman.GetTryNewOutboundPeer()) {
+                m_connman.SetTryNewOutboundPeer(false);
+            }
+
+            // Update connections desirable services flags in case we need to. For instance,
+            // the tip could be lagging behind, further than the limited peers threshold.
+            MaybeUpdateDesirableConnOptions();
         }
+
         m_stale_tip_check_time = now + STALE_CHECK_INTERVAL;
     }
```
```

Reply:
Sure, thanks!. Done as suggested.